plugins/util/portmap/portmap.go

```diff
@@ -0,0 +1,223 @@
+// Copyright 2017 CNI authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"fmt"
+	"net"
+	"strconv"
+
+	"github.com/containernetworking/cni/pkg/types/current"
+	"github.com/containernetworking/cni/pkg/utils"
+	"github.com/containernetworking/cni/pkg/utils/sysctl"
+	"github.com/coreos/go-iptables/iptables"
+	"github.com/vishvananda/netlink"
+)
+
+// forwardPorts creates iptables rules to forward traffic from ports on
+// If enableSNAT is set, it also creates a masquerading chain that rewrites
+// connections to localhost:<port> so they also work.
+func forwardPorts(config *PortMapConf, containerIP net.IP, enableSNAT bool) error {
+	isV6 := (containerIP.To4() == nil)
+
+	var ipt *iptables.IPTables
+	var err error
+
+	if isV6 {
+		ipt, err = iptables.NewWithProtocol(iptables.ProtocolIPv6)
+	} else {
+		ipt, err = iptables.NewWithProtocol(iptables.ProtocolIPv4)
+	}
+	if err != nil {
+		return fmt.Errorf("failed to open iptables:", err)
+	}
+
+	dnatChain := genDnatChain(config.Name, config.ContainerID)
+	dnatRules := dnatRules(config.RuntimeConfig.PortMaps, containerIP)
+	if err := dnatChain.setup(ipt, dnatRules); err != nil {
+		return fmt.Errorf("unable to setup DNAT", err)
+	}
+
+	if enableSNAT {
+		snatChain := genSnatChain(config.Name, config.ContainerID, isV6)
+		snatRules := snatRules(config.RuntimeConfig.PortMaps, containerIP)
+		if err := snatChain.setup(ipt, snatRules); err != nil {
+			return fmt.Errorf("unable to setup SNAT", err)
+		}
+		if !isV6 {
+			hostIfName := getRoutableHostIF(config.PrevResult)
+			if hostIfName != "" {
+				if err := enableLocalnetRouting(hostIfName); err != nil {
+					return fmt.Errorf("unable to enable route_localnet: %v", err)
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+func genDnatChain(netName, containerID string) chain {
+	name := utils.FormatChainName("DN-", netName, containerID)
+	comment := fmt.Sprintf(`dnat name: "%s" id: "%s"`, netName, containerID)
+
+	return chain{
+		name: name,
+		entryRule: []string{
+			"-m", "addrtype",
+			"--dst-type", "LOCAL",
+			"-m", "comment",
+			"--comment", comment,
+			"-j", name},
+		entryChains: []string{"PREROUTING", "OUTPUT"},
```

Comment:
I am not sure whether most OS distros have it by default. Since the number of hostport on a node will not be too high, I think it may not worth adding this dependency.